README.md
# mc-user
[![Latest Release](https://img.shields.io/github/v/release/Ubunfu/mc-user)](https://github.com/Ubunfu/mc-user/releases)
[![codecov](https://codecov.io/gh/Ubunfu/mc-user/branch/master/graph/badge.svg?token=41OYQX1DGV)](https://codecov.io/gh/Ubunfu/mc-user)
[![CircleCI](https://img.shields.io/circleci/build/github/Ubunfu/mc-user?logo=circleci)](https://app.circleci.com/pipelines/github/Ubunfu/mc-user)
![Contrubutors](https://img.shields.io/github/contributors/Ubunfu/mc-user?color=blue)
![Last Commit](https://img.shields.io/github/last-commit/Ubunfu/mc-user)

Maintains and provides access to a correlation between Discord usernames and Minecraft usernames.

This service runs as an AWS lambda function.

## Configuration
### IAM Role
AWS's standard IAM role for Lambda micro services is plenty sufficient. The only access that is required is read/write for DynamoDB, and write to CloudWatch logs.

* AWSLambdaMicroserviceExecutionRole
* AWSLambdaBasicExecutionRole

### Environment Variables
| Parameter                    | Description                                                                       | Default | Required? |
|------------------------------|-----------------------------------------------------------------------------------|---------|-----------|
| TABLE_USERS                  | The name of a DynamoDB table correlating Discord usernames to Minecraft usernames.| n/a     | Yes       |
| LOGGER_ENABLED               | Boolean value controlling writing of logs. Useful to turn off for test execution. | n/a     | Yes       |

CHANGELOG.md
# Changelog
This project attempts to conform to [SemVer](https://semver.org/) guidelines whenever possible.

## [v1.0.0]
* [Adding db to serverless stack](https://github.com/Ubunfu/mc-user/pull/5)
* [Configuring discord notifications from CircleCI](https://github.com/Ubunfu/mc-user/pull/6)

## [v0.2.1]
* [Fixing broken function name in serverless config](https://github.com/Ubunfu/mc-user/pull/4)

## [v0.2.0]
* [Adopting serverless framework](https://github.com/Ubunfu/mc-user/pull/2)
* [Adding UML diagrams](https://github.com/Ubunfu/mc-user/pull/1)

## [v0.1.0]
* Initial implementation

test/indexTest.js
const index = require('../src/index');
const userService = require('../src/service/userService');
const sinon = require('sinon');
const { expect } = require('chai');

const EVENT_GET_USER = {
    requestContext: {
        routeKey: 'GET /user'
    },
    queryStringParameters: {
        discordUser: 'discordUsername'
    }
}

const EVENT_UNKNOWN_REQ = {
    requestContext: {
        routeKey: 'GET /unknown/api'
    }
}

const RESP_ERROR_USER_NOT_FOUND = {
    error: 'get user failed',
    errorDetail: 'user not found'
}
const RESP_ERROR_OTHER = {
    error: 'get user failed',
    errorDetail: 'anything else'
}

const A_USER = {
    discordUser: 'discordUsername',
    minecraftUser: 'minecraftUsername'
}

describe('index: When unknown request is received', function() {
    it('Should return HTTP 404', async function() {
        const indexResp = await index.handler(EVENT_UNKNOWN_REQ);
        expect(indexResp.statusCode).to.be.equal('404');
    });
});

describe('index: When GET user API is called', function() {
    describe('And DB returns user data', function() {
        it('Returns HTTP 200', async function() {
            const userServiceMock = sinon.stub(userService, "getUserByDiscordId").returns(A_USER)
            const indexResp = await index.handler(EVENT_GET_USER)
            expect(indexResp.statusCode).to.be.equal('200')
            expect(JSON.parse(indexResp.body)).to.be.deep.equal(A_USER)
            userServiceMock.restore()
        })
    })
    describe('And DB does not find a matching user', function() {
        it('Returns HTTP 404', async function() {
            const userServiceMock = sinon.stub(userService, "getUserByDiscordId")
                .throws('errorName', 'user not found')
            const indexResp = await index.handler(EVENT_GET_USER)
            expect(indexResp.statusCode).to.be.equal('404')
            expect(JSON.parse(indexResp.body)).to.be.deep.equal(RESP_ERROR_USER_NOT_FOUND)
            userServiceMock.restore()
        })
    })
    describe('And error occurs checking the DB', function() {
        it('Returns HTTP 500', async function() {
            const userServiceMock = sinon.stub(userService, "getUserByDiscordId")
                .throws('errorName', 'anything else')
            const indexResp = await index.handler(EVENT_GET_USER)
            expect(indexResp.statusCode).to.be.equal('500')
            expect(JSON.parse(indexResp.body)).to.be.deep.equal(RESP_ERROR_OTHER)
            userServiceMock.restore()
        })
    })
})

test/service/userServiceTest.js
const { expect } = require('chai');
const sinon = require('sinon');
const userService = require('../../src/service/userService');

const A_USER = {
    discordUser: 'discordUsername',
    minecraftUser: 'minecraftUsername'
}

const DB_RESP_USER = {
    Item: A_USER
}

describe('userService: When getUserByDiscordId is called', function() {
    describe('And DB returns a user', function() {
        it('Returns the user', async function() {
            const fakeDocClient = {
                get: sinon.stub().returnsThis(),
                promise: sinon.stub().resolves(DB_RESP_USER)
            }
            const item = await userService.getUserByDiscordId(fakeDocClient, 'discordUsername')
            expect(item).to.be.deep.equal(A_USER)
        })
    })
    describe('And DB does not find a matching user', function() {
        it('Throws a \'user not found\' error', async function() {
            const fakeDocClient = {
                get: sinon.stub().returnsThis(),
                promise: sinon.stub().resolves({})
            }
            try {
                await userService.getUserByDiscordId(fakeDocClient, 'discordUsername')
                expect(true).to.be.false
            } catch (err) {
                expect(err.message).to.be.equal('user not found')
            }
        })
    })
    describe('And an error is thown while connecting to the DB', function() {
        it('Throws the error back up to the caller', async function() {
            const fakeDocClient = {
                get: sinon.stub().returnsThis(),
                promise: sinon.stub().rejects('errorName', 'some error')
            }
            try {
                await userService.getUserByDiscordId(fakeDocClient, 'discordUsername')
                expect(true).to.be.false
            } catch (err) {
                expect(err.message).to.be.equal('some error')
            }
        })
    })
})

src/util/rconClearResponseProcessor.js
const itemService = require('../service/itemService.js');
const walletService = require('../service/walletService.js');
const axios = require('axios');
const AWS = require('aws-sdk');
const docClient = new AWS.DynamoDB.DocumentClient();
const { log } = require('./logger.js');

const SPACE = ' ';

async function processClearResponse(resp, itemName) {
    if (resp.includes('Removed')) {
        await processRemovedItems(resp, itemName);
    } else if (resp.includes('No items were found on player'))  {
        await processNoItemsFound(resp, itemName);
    } else if (resp.includes('No player was found')) {
        await processNoPlayerFound();
    } else {
        await processUnknownResponse(resp);
    }
}

async function processRemovedItems(resp, itemName) {
    let item;
    try {
        item = await itemService.getItem(docClient, itemName);
    } catch (err) {
        log('[rconClearResponseProcessor] Error getting item: ' + err.message);
        throw Error('Error getting item from DB');
    }

    const words = resp.split(SPACE);
    const playerName = words[5];
    const itemQuantity = words[1];
    const totalPrice = item.sellPrice * itemQuantity;

    try {
        await walletService.payWallet(playerName, totalPrice);
    } catch (err) {
        log('[rconClearResponseProcessor] Error paying wallet: ' + err.message);
        throw Error('Error paying wallet');
    }
    
    try {
        await postSaleToWebhook(playerName, itemName, itemQuantity, totalPrice);
    } catch (err) {
        log('[rconClearResponseProcessor] Error posting to Discord: ' + err.message);
        throw Error('Error posting notification to Discord');
    }
}

async function processNoItemsFound(resp, itemName) {
    
    const words = resp.split(SPACE);
    const playerName = words[6];

    try {
        await postNoItemsFoundToWebhook(itemName, playerName);
    } catch (err) {
        log('[rconClearResponseProcessor] Error posting to Discord: ' + err.message);
        throw Error('Error posting notification to Discord');
    }
}

async function processNoPlayerFound() {
    try {
        await postNoPlayerFoundToWebhook();
    } catch (err) {
        log('[rconClearResponseProcessor] Error posting to Discord: ' + err.message);
        throw Error('Error posting notification to Discord');
    }
}

async function processUnknownResponse(resp) {
    try {
        await postUnknownResponseToWebhook(resp);
    } catch (err) {
        log('[rconClearResponseProcessor] Error posting to Discord: ' + err.message);
        throw Error('Error posting notification to Discord');
    }
}

async function postSaleToWebhook(playerName, itemName, itemQuantity, totalPrice) {
    await axios.post(
        process.env.DISCORD_WEBHOOK_URL,
        {
            "embeds": [
                {
                    "author": {
                        "name": "Wandering Trader"
                    },
                    "thumbnail": {
                        "url": process.env.DISCORD_WEBHOOK_THUMBNAIL_URL
                    },
                    "title": `Purchased ${itemQuantity} ${itemName} from ${playerName} for ${totalPrice}`,
                    "color": 65280
                }
            ]
        }
    );
}

async function postNoItemsFoundToWebhook(itemName, playerName) {
    await axios.post(
        process.env.DISCORD_WEBHOOK_URL,
        {
            "embeds": [
                {
                    "author": {
                        "name": "Wandering Trader"
                    },
                    "thumbnail": {
                        "url": process.env.DISCORD_WEBHOOK_THUMBNAIL_URL
                    },
                    "title": `No ${itemName} found in ${playerName} inventory.  Sale aborted.`,
                    "color": 65280
                }
            ]
        }
    );
}

async function postNoPlayerFoundToWebhook() {
    await axios.post(
        process.env.DISCORD_WEBHOOK_URL,
        {
            "embeds": [
                {
                    "author": {
                        "name": "Wandering Trader"
                    },
                    "thumbnail": {
                        "url": process.env.DISCORD_WEBHOOK_THUMBNAIL_URL
                    },
                    "title": `Player not online.  Sale aborted.`,
                    "color": 65280
                }
            ]
        }
    );
}

async function postUnknownResponseToWebhook(resp) {
    await axios.post(
        process.env.DISCORD_WEBHOOK_URL,
        {
            "embeds": [
                {
                    "author": {
                        "name": "Wandering Trader"
                    },
                    "thumbnail": {
                        "url": process.env.DISCORD_WEBHOOK_THUMBNAIL_URL
                    },
                    "title": `I tried to process the transaction, but the server responded unexpectedly: \`\`\`${resp}\`\`\`.`,
                    "color": 65280
                }
            ]
        }
    );
}

exports.processClearResponse = processClearResponse;

src/index.js
require('dotenv').config();
const { log } = require('./util/logger.js');
const userService = require('./service/userService');
const AWS = require('aws-sdk');
const docClient = new AWS.DynamoDB.DocumentClient();

exports.handler = async (event, context) => {
    await log('Received event: ' + JSON.stringify(event, null, 2));

    let body;
    let statusCode = '200';
    const headers = {
        'Content-Type': 'application/json',
    };

    if (event.requestContext.routeKey == 'GET /user') {
        const discordUser = event.queryStringParameters.discordUser;
        try {
            body = await userService.getUserByDiscordId(docClient, discordUser)
        } catch (err) {
            if (err.message == 'user not found') {
                statusCode = '404';
            } else {
                statusCode = '500';
            }
            body = {
                error: 'get user failed',
                errorDetail: err.message
            }
        }
    } else {
        statusCode = '404'
    }

    body = JSON.stringify(body);
    return {
        statusCode,
        body,
        headers,
    };
};